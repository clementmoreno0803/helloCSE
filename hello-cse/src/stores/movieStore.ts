import { movieDetail } from "@/models/movieDetailInterface";
import { topMovieInterface } from "@/models/topMovieInterface";
import { upComingMoviesInterface } from "@/models/upCommingMovieInterface";
import { defineStore } from "pinia";
import { computed, ref } from "vue";
import { currentMovieInterface } from "@/models/currentMovieInterface";

export const useMovieStore = defineStore("movieStore", () => {
  const topMovies = ref<topMovieInterface[]>([]);
  const currentMovies = ref<currentMovieInterface[]>([]);
  const upComingMovies = ref<upComingMoviesInterface[]>([]);
  const movieDetails = ref<movieDetail>();
  const inputFilter = ref<string>("");
  const isLoaded = ref<boolean>(false);

  //getter

  const getFilteredMovies = computed(() => {
    return upComingMovies.value.filter((movie) => movie.title.toLowerCase().includes(inputFilter.value.toLowerCase()));
  });

  const getTopFiveMovies = computed(() => {
    const moviesSorted = topMovies.value.sort((a, b) => b.vote_average - a.vote_average);
    return moviesSorted.slice(0, 5);
  });

  const getHeightCurrentMovies = computed(() => {
    return currentMovies.value.slice(0, 8);
  });

  const listTop4Actors = computed(() => {
    const castingSorted = movieDetails.value?.cast.sort((a, b) => b.popularity - a.popularity);
    return castingSorted?.slice(0, 4);
  });

  //setter
  const setTopMovies = (newMovies: topMovieInterface[]) => {
    topMovies.value = newMovies;
  };
  const setCurrentMovies = (newCurrentMovie: currentMovieInterface[]) => {
    currentMovies.value = newCurrentMovie;
  };
  const setUpComingMovies = (newMovies: upComingMoviesInterface[]) => {
    upComingMovies.value = newMovies;
    isLoaded.value = true;
  };
  const setMovieFilterByName = (input: string) => {
    inputFilter.value = input;
  };
  const setMovieDetails = (details: movieDetail | undefined) => {
    if (details) {
      return (movieDetails.value = details);
    }
  };

  return {
    topMovies,
    upComingMovies,
    inputFilter,
    getFilteredMovies,
    getTopFiveMovies,
    getHeightCurrentMovies,
    movieDetails,
    listTop4Actors,
    isLoaded,
    setTopMovies,
    setCurrentMovies,
    setUpComingMovies,
    setMovieFilterByName,
    setMovieDetails
  };
});
