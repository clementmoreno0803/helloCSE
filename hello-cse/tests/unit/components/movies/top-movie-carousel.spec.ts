import { shallowMount } from '@vue/test-utils';
import { describe, it, expect, vi } from 'vitest';
import { createTestingPinia } from '@pinia/testing';
import { useMovieStore } from '@/stores/movieStore';
import { createRouter, createWebHistory } from "vue-router";
import TopMovieCarousel from "@/components/movies/top-movie-carousel.vue";
import { mockMovies } from "../../fixtures/mockMovies.mock";

const router = createRouter({
  history: createWebHistory(),
  routes: [
    { path: '/', component: TopMovieCarousel },
  ],
});


  const generateWrapper = () => {
    return shallowMount(TopMovieCarousel, {
      global: {
        plugins: [
          router,
          createTestingPinia({
            initialState: {
              movieStore: {
                getTopFiveMovies: [],
              },
            },
            createSpy: vi.fn,
          }),
        ],
      },
    });
  };

describe('TopMovieCard.vue', () => {
  it('should show skeleton loaders when no movies are available', async () => {
    const wrapper = generateWrapper();

    const skeletonLoaders = wrapper.findAll('v-skeleton-loader');

    expect(skeletonLoaders.length).toBe(2);
  });

  it('should display top 5 movies when available', async () => {
    const wrapper = generateWrapper();
    useMovieStore().getTopFiveMovies = mockMovies
    await wrapper.vm.$nextTick()

    const movieItems = wrapper.findAll('.top-movie-card__carousel-item');
    expect(movieItems.length).toBe(5);
    expect(movieItems[0].text()).toContain(mockMovies[0].title);
    expect(movieItems[0].text()).toContain(mockMovies[0].vote_average.toString());
  });
});
