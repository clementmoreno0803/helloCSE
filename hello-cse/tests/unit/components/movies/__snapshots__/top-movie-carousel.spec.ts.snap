// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TopMovieCard.vue > should render correctly 1`] = `
<div
  data-v-5a6928d2=""
>
  <v-skeleton-loader
    class="top-movie-card__title"
    color="rgb(28,28,28)"
    data-v-5a6928d2=""
    elevation="3"
    type="paragraph"
  />
  <div
    data-v-5a6928d2=""
  >
    <v-skeleton-loader
      class="top-movie-card__carousel-container"
      color="rgb(28,28,28)"
      data-v-5a6928d2=""
      elevation="3"
      type="card"
    />
  </div>
</div>
`;
